src/compute/src/render/context.rs

```diff
@@ -212,27 +218,305 @@ impl ShutdownToken {
     }
 }
 
+/// A representation of arrangements that are statically type-specialized.
+/// Each variant of this `enum` covers a different supported specialization of
+/// key and value types for arrangement flavors.
+///
+/// The specialization here is performed on the representation length, as opposed
+/// to its constituent types. For fixed-length specializations, it thus becomes
+/// necessary to keep track of the schema used, since datums are not used and thus
+/// the representation is not tagged. A catch-all `RowRow` specialization without
+/// schema information allows for covering the current approach of self-describing
+/// variable-length keys and variable-length values.
+#[derive(Clone)]
+pub enum SpecializedArrangement<S: Scope>
+where
+    <S as ScopeParent>::Timestamp: Lattice,
+{
+    Bytes9Row(Vec<ColumnType>, KeyValArrangement<S, Bytes9, Row>),
+    RowRow(KeyValArrangement<S, Row, Row>),
+}
+
+impl<S: Scope> SpecializedArrangement<S>
+where
+    <S as ScopeParent>::Timestamp: Lattice,
+{
+    /// The scope of the underlying arrangement's stream.
+    pub fn stream_scope(&self) -> S {
+        match self {
+            SpecializedArrangement::Bytes9Row(_, inner) => inner.stream.scope(),
+            SpecializedArrangement::RowRow(inner) => inner.stream.scope(),
+        }
+    }
+
+    /// Brings the underlying arrangement into a region.
+    pub fn enter_region<'a>(
+        &self,
+        region: &Child<'a, S, S::Timestamp>,
+    ) -> SpecializedArrangement<Child<'a, S, S::Timestamp>> {
+        match self {
+            SpecializedArrangement::Bytes9Row(key_types, inner) => {
+                SpecializedArrangement::Bytes9Row(key_types.clone(), inner.enter_region(region))
+            }
+            SpecializedArrangement::RowRow(inner) => {
+                SpecializedArrangement::RowRow(inner.enter_region(region))
+            }
+        }
+    }
+
+    /// Extracts the underlying arrangement as a stream of updates.
+    pub fn as_collection<L>(&self, mut logic: L) -> Collection<S, Row, Diff>
+    where
+        L: FnMut(&Row, &Row) -> Row + 'static,
+    {
```

Comment:
The return type seems surprising, especially because there is only one callsite that immediately decodes the row. It'd be nice to avoid this round-trip step, but it might be hard due to lifetimes.

Reply:
A similar solution to what was implemented for `flat_map` works here as well. Incorporated now.